service/dap/server.go

```diff
@@ -2570,6 +2575,49 @@ func (s *Server) resetHandlesForStoppedEvent() {
 	s.exceptionErr = nil
 }
 
+func (s *Server) setRunning(running bool) {
+	s.runningMutex.Lock()
+	s.running = running
+	s.runningMutex.Unlock()
+}
+
+func (s *Server) IsRunning() bool {
+	s.runningMutex.Lock()
+	defer s.runningMutex.Unlock()
+	return s.running
+}
+
+// halt is a helper function to allow dap to make sure that a halt
+// is processed. It is possible that this halt request will come in while a
+// a breakpoint is being processed. If the debugger decides to resume execution
+// after processing the breakpoint, the halt request would be skipped. Additional
+// synchronization is required between "resume" and "halt" to make sure this does
+// not happen.
+func (s *Server) halt() (*api.DebuggerState, error) {
+	s.stopMutex.Lock()
+	defer s.stopMutex.Unlock()
+	s.setRunning(false)
```

Comment:
I think the running flag should be fully controlled by the doRunCommand goroutine. Between the point where this line is executed and the point where doRunCommand finishes the value of the running flag is lying. Maybe it doesn't have any adverse effects but it bothers me.

Reply:
I think that makes sense.

If we keep the StoppedReason to not update after the fact, then we would need to call CheckAndClearManualHalt() from the dap code to check if a manual halt was issued OR add another tracking mechanism within DAP server to check if a halt was issued. I think the second would be too many different things to track when the target is already tracking it.

Reply:
I think tracking it explicitly is the better option, someone reading dap/server.go will not have to understand that StopReason always reports StopManual if a manual stop request was issued, especially since we make no such guarantee in the documentation.